packages/gatsby/src/commands/develop.js

```diff
@@ -195,33 +191,188 @@ async function startServer(program) {
+            )
+          ) {
+            // Address is private, format it for later use
+            lanUrlForTerminal = prettyPrintUrl(lanUrlForConfig)
+          } else {
+            // Address is not private, so we will discard it
+            lanUrlForConfig = undefined
+          }
+        }
+      } catch (_e) {
+        // ignored
+      }
+    } else {
+      prettyHost = host
+    }
+    // TODO collect errors (GraphQL + Webpack) in Redux so we
+    // can clear terminal and print them out on every compile.
+    // Borrow pretty printing code from webpack plugin.
+    const localUrlForTerminal = prettyPrintUrl(prettyHost)
+    const localUrlForBrowser = formatUrl(prettyHost)
+    return {
+      lanUrlForConfig,
+      lanUrlForTerminal,
+      localUrlForTerminal,
+      localUrlForBrowser,
+    }
+  }
+
+  function printInstructions(appName, urls, useYarn) {
+    console.log()
+    console.log(`You can now view ${chalk.bold(appName)} in the browser.`)
+    console.log()
+
+    if (urls.lanUrlForTerminal) {
+      console.log(
+        `  ${chalk.bold(`Local:`)}            ${urls.localUrlForTerminal}`
+      )
+      console.log(
+        `  ${chalk.bold(`On Your Network:`)}  ${urls.lanUrlForTerminal}`
+      )
+    } else {
+      console.log(`  ${urls.localUrlForTerminal}`)
     }
 
-    return startServer(program)
+    console.log()
+    console.log(`Note that the development build is not optimized.`)
+    console.log(
+      `To create a production build, use ` + `${chalk.cyan(`gatsby build`)}`
+    )
+    console.log()
+  }
+
+  const host =
+    listener.address().address === `127.0.0.1`
+      ? `localhost`
+      : listener.address().address
+
+  let isFirstCompile = true
+  // "done" event fires when Webpack has finished recompiling the bundle.
+  // Whether or not you have warnings or errors, you will get this event.
+  compiler.plugin(`done`, stats => {
+    // We have switched off the default Webpack output in WebpackDevServer
+    // options so we are going to "massage" the warnings and errors and present
+    // them in a readable focused way.
+    const messages = formatWebpackMessages(stats.toJson({}, true))
+    const urls = prepareUrls(`http`, host, port)
+    const isSuccessful = !messages.errors.length && !messages.warnings.length
+    // if (isSuccessful) {
+    // console.log(chalk.green(`Compiled successfully!`))
+    // }
+    // if (isSuccessful && (isInteractive || isFirstCompile)) {
+    if (isSuccessful && isFirstCompile) {
+      printInstructions(program.appName, urls, program.useYarn)
+    }
+
+    if (program.open) {
+      require(`opn`)(urls.localUrlForBrowser)
+    }
+
+    isFirstCompile = false
+
+    // If errors exist, only show errors.
+    // if (messages.errors.length) {
+    // // Only keep the first error. Others are often indicative
+    // // of the same problem, but confuse the reader with noise.
+    // if (messages.errors.length > 1) {
+    // messages.errors.length = 1
```

Comment:
we'll restore this replacing `FriendlyErrorsWebpackPlugin` once the remainder of the error/warning refactoring is done.